stdlib/mqtt/to.go

```diff
@@ -0,0 +1,649 @@
+						m.name = er.Strings(j).ValueString(i)
+					case isTag[j]:
+						if col.Type != flux.TString {
+							return errors.New("invalid type for tag column")
+						}
+						m.tags = append(m.tags, &protocol.Tag{Key: col.Label, Value: er.Strings(j).ValueString(i)})
+
+					case isValue[j]:
+						switch col.Type {
+						case flux.TFloat:
+							m.fields = append(m.fields, &protocol.Field{Key: col.Label, Value: er.Floats(j).Value(i)})
+						case flux.TInt:
+							m.fields = append(m.fields, &protocol.Field{Key: col.Label, Value: er.Ints(j).Value(i)})
+						case flux.TUInt:
+							m.fields = append(m.fields, &protocol.Field{Key: col.Label, Value: er.UInts(j).Value(i)})
+						case flux.TString:
+							m.fields = append(m.fields, &protocol.Field{Key: col.Label, Value: er.Strings(j).ValueString(i)})
+						case flux.TTime:
+							m.fields = append(m.fields, &protocol.Field{Key: col.Label, Value: values.Time(er.Times(j).Value(i))})
+						case flux.TBool:
+							m.fields = append(m.fields, &protocol.Field{Key: col.Label, Value: er.Bools(j).Value(i)})
+						default:
+							return fmt.Errorf("invalid type for column %s", col.Label)
+						}
+					}
+				}
+				_, err := e.Encode(m)
+				if err != nil {
+					return err
+				}
+
+				if err := execute.AppendRecord(i, er, builder); err != nil {
+					return err
+				}
+			}
+			return nil
+		})
+		if e := pw.Close(); e != nil && err == nil {
+			err = e
+		}
+		return err
+	})
+	
+//start a client using the above ClientOptions
+	if token := client.Connect(); token.Wait() && token.Error() != nil {
+		return token.Error()
+	}
+	p := make([]byte, 2024) 
+	var message strings.Builder
+	lines := 0
+	// messages come in as triples: measurement & tags, values, timestamp
+	// put them all together and you have line-protocol! 
+	for {
+		n, err := pr.Read(p)
+		if err != nil{
+		    if err == io.EOF {
+				message.WriteString(string(p[:n]))
+				break
+			}
+			client.Disconnect(250)
+			return err
+		}
+		message.WriteString(string(p[:n])) // handle leftovers
+		lines += 1
+		if lines > 2 { // post message after full row is read.
+			if len(mqttTopic) <= 0 {// No topic set? Create topic out of tags
+				mqttTopic = m.createTopic(message.String())				
+			}
+			if t.spec.Spec.Format == "JSON" { // format message as a JSON
+				message = m.formatJSON(message.String())
+			}
+			token := client.Publish(mqttTopic, 0, false, message.String())
+			token.Wait()
+			lines = 0
+			message.Reset()
+		}
+	}	
+	if len(message.String()) > 0 { // if any leftover messages are there, write them out. 
+		if len(mqttTopic) <= 0 {// create topic out of tags
+			mqttTopic = m.createTopic(message.String())				
+		}
+		token := client.Publish(mqttTopic, 0, false, message.String())
+	  	token.Wait()
+	}
+	if err := wg.Wait(); err != nil {
+		client.Disconnect(250)
+		return err
+	}
+	client.Disconnect(250)
+	return nil
+
+}
+// format the message as json
+func (t *toMqttMetric) formatJSON(message string) strings.Builder {
```

Comment:
this is sort of an antipattern in Go.  You can use struct annotations to marshal a simple struct as JSON.   so you have something like   `type metric struct{...}` that can be encoded properly as JSON, and then you can have a `func (m metric) toMessage() string {...}`   to encode the message as a simple string.